x/wasm/internal/keeper/recurse_test.go

```diff
@@ -0,0 +1,259 @@
+	Hashed []byte `json:"hashed"`
+}
+
+func TestGasCostOnQuery(t *testing.T) {
+	const (
+		GasNoWork uint64 = InstanceCost + 2_756
+		// Note: about 100 SDK gas (10k wasmer gas) for each round of sha256
+		GasWork50 uint64 = InstanceCost + 8_464 // this is a little shy of 50k gas - to keep an eye on the limit
+
+		GasReturnUnhashed uint64 = 647
+		GasReturnHashed   uint64 = 597
+	)
+
+	cases := map[string]struct {
+		gasLimit    uint64
+		msg         Recurse
+		expectedGas uint64
+	}{
+		"no recursion, no work": {
+			gasLimit:    400_000,
+			msg:         Recurse{},
+			expectedGas: GasNoWork,
+		},
+		"no recursion, some work": {
+			gasLimit: 400_000,
+			msg: Recurse{
+				Work: 50, // 50 rounds of sha256 inside the contract
+			},
+			expectedGas: GasWork50,
+		},
+		"recursion 1, no work": {
+			gasLimit: 400_000,
+			msg: Recurse{
+				Depth: 1,
+			},
+			expectedGas: 2*GasNoWork + GasReturnUnhashed,
+		},
+		"recursion 1, some work": {
+			gasLimit: 400_000,
+			msg: Recurse{
+				Depth: 1,
+				Work:  50,
+			},
+			expectedGas: 2*GasWork50 + GasReturnHashed,
+		},
+		"recursion 4, some work": {
+			gasLimit: 400_000,
+			msg: Recurse{
+				Depth: 4,
+				Work:  50,
+			},
+			// this is (currently) 244_708 gas
+			expectedGas: 5*GasWork50 + 4*GasReturnHashed,
+		},
+	}
+
+	// we do one basic setup before all test cases (which are read-only and don't change state)
+	tempDir, err := ioutil.TempDir("", "wasm")
+	require.NoError(t, err)
+	defer os.RemoveAll(tempDir)
+
+	ctx, keepers := CreateTestInput(t, false, tempDir, SupportedFeatures, nil, nil)
+	accKeeper, keeper := keepers.AccountKeeper, keepers.WasmKeeper
+	deposit := sdk.NewCoins(sdk.NewInt64Coin("denom", 100000))
+	creator := createFakeFundedAccount(ctx, accKeeper, deposit.Add(deposit...))
+
+	// store the code
+	wasmCode, err := ioutil.ReadFile("./testdata/contract.wasm")
+	require.NoError(t, err)
+	codeID, err := keeper.Create(ctx, creator, wasmCode, "", "", nil)
+	require.NoError(t, err)
+
+	// instantiate the contract
+	_, _, bob := keyPubAddr()
+	_, _, fred := keyPubAddr()
+	initMsg := InitMsg{
+		Verifier:    fred,
+		Beneficiary: bob,
+	}
+	initMsgBz, err := json.Marshal(initMsg)
+	require.NoError(t, err)
+	contractAddr, err := keeper.Instantiate(ctx, codeID, creator, nil, initMsgBz, "recursive contract", deposit)
+	require.NoError(t, err)
+
+	for name, tc := range cases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			// external limit has no effect (we get a panic if this is enforced)
+			keeper.queryGasLimit = 1000
+
+			// make sure we set a limit before calling
+			ctx = ctx.WithGasMeter(sdk.NewGasMeter(tc.gasLimit))
+			require.Equal(t, uint64(0), ctx.GasMeter().GasConsumed())
+
+			recurse := tc.msg
+			recurse.Contract = contractAddr
+			msg := buildQuery(t, recurse)
+
+			// this should throw out of gas exception (panic)
+			data, err := keeper.QuerySmart(ctx, contractAddr, msg)
+			require.NoError(t, err)
+			var resp recurseResponse
```

Comment:
pp: As you testing for the gas cost the response value does not really contribute to this goal. I would remove L135-146 to stay focused here

Reply:
I want to make sure the right stuff is coming out, but I could do that in another test.
I will remove the check for the higher-level external queries